File: src/main/java/com/example/progettoenterprise/data/service/SegnalazioneService.java
package com.example.progettoenterprise.data.service;
import com.example.progettoenterprise.data.repositories.specifications.SegnalazioneSpecification;
import com.example.progettoenterprise.dto.SegnalazioneDTO;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;

public interface SegnalazioneService {
    SegnalazioneDTO creaSegnalazione(SegnalazioneDTO segnalazioneDTO, Long idSegnalatore);
    org.springframework.data.domain.Page<SegnalazioneDTO> cercaSegnalazioni(SegnalazioneSpecification.SegnalazioneFilter filtro, int pagina, int dimensione);
    SegnalazioneDTO prendiInCarico(Long segnalazioneId, Long adminId);
    SegnalazioneDTO risolviSegnalazione(Long idSegnalazione, Long idAdmin, boolean sospendiAutore);

    SegnalazioneDTO rifiutaSegnalazione(Long segnalazioneId, Long adminId);
    long contaSegnalazioniAperte();
}

File: frontendWeb/src/app/service/segnalazione.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';

@Injectable({
  providedIn: 'root'
})
export class SegnalazioneService {

  private urlBase = '/api/v1/segnalazioni';

  constructor(private http: HttpClient) {}

  creaSegnalazione(datiSegnalazione: any, idSegnalatore: number): Observable<any> {
    return this.http.post(`${this.urlBase}/crea?idSegnalatore=${idSegnalatore}`, datiSegnalazione);
  }

  cercaSegnalazioni(filtro: any, pagina: number, dimensione: number = 10): Observable<any> {
    let params = new HttpParams()
      .set('pagina', pagina.toString())
      .set('dimensione', dimensione.toString());

    // Aggiungi i filtri solo se esistono
    if (filtro.tipo) {
      params = params.set('tipo', filtro.tipo);
    }
    if (filtro.stato) {
      params = params.set('stato', filtro.stato);
    }
    if (filtro.segnalatoreId) {
      params = params.set('segnalatoreId', filtro.segnalatoreId.toString());
    }
    if (filtro.adminId) {
      params = params.set('adminId', filtro.adminId.toString());
    }

    return this.http.get<any>(`${this.urlBase}/ricerca`, { params });
  }
  prendiInCarico(idSegnalazione: number, idAdmin: number): Observable<any> {
    return this.http.put(`${this.urlBase}/${idSegnalazione}/prendi-in-carico?idAdmin=${idAdmin}`, {});
  }

  risolviSegnalazione(idSegnalazione: number, idAdmin: number, sospendiAutore: boolean = false): Observable<any> {
    return this.http.put(`${this.urlBase}/${idSegnalazione}/risolvi?idAdmin=${idAdmin}&sospendiAutore=${sospendiAutore}`, {});
  }

  rifiutaSegnalazione(idSegnalazione: number, idAdmin: number): Observable<any> {
    return this.http.put(`${this.urlBase}/${idSegnalazione}/rifiuta?idAdmin=${idAdmin}`, {});
  }

  contaSegnalazioniAperte(): Observable<number> {
    return this.http.get<number>(`${this.urlBase}/contatore-aperte`);
  }
}
